### DISCUSSION TITLE
[Doc] "negatively acknowledged cumulatively" ambiguous?

### DISCUSSION THREAD

**Discussion by startjava** [apache/pulsar#18771] [General]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### What issue do you find in Pulsar docs?

This website:

https://pulsar.apache.org/docs/concepts-messaging/#negative-acknowledgement

The:

Messages are negatively acknowledged individually or cumulatively, depending on the consumption subscription type.

The inner "opened indispensable cumulatively", does it have this function in pulsar? At least I did not find the corresponding API for this function.

I don't know if there is some ambiguity in the document, or if there is some problem with my understanding.

In addition:

In Exclusive and Failover subscription types, consumers only negatively acknowledge the last message they receive.

In Shared and Key_Shared subscription types, consumers can negatively acknowledge messages individually.

Are these two paragraphs necessary? In fact, it is the feature of four subscription types, namely whether there is one consumer or how many consumers, and it feels like adding these two paragraphs adds complexity to this technical point.

 

### What is your suggestion?

?

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by startjava** [comment]
Actually I want to know what situation belongs to "the acknowledged cumulatively"?
thank you !

**Comment by zymap** [comment]
Do you want this? https://pulsar.apache.org/docs/concepts-messaging#acknowledgement

**Comment by startjava** [comment]
> Do you want this? https://pulsar.apache.org/docs/concepts-messaging#acknowledgement

YES，i see the doc too .

**Comment by startjava** [comment]
@momo-jun please see the question

**Comment by momo-jun** [comment]
> Messages are negatively acknowledged individually or cumulatively, depending on the consumption subscription type.
> * In Exclusive and Failover subscription types, consumers only negatively acknowledge the last message they receive.
> * In Shared and Key_Shared subscription types, consumers can negatively acknowledge messages individually.

I don't see the ambiguity in this paragraph. It means the negative acknowledgment of messages can be performed either individually (in Shared and Key_Shared subscription types) or cumulatively (Exclusive and Failover subscription types).

Did this answer your question? @startjava

**Comment by startjava** [comment]
@momo-jun 

question1：
why 
（1）individually (in Shared and Key_Shared subscription types)
（2）cumulatively (Exclusive and Failover subscription types).

why bottom is error：
（1）cumulatively  (in Shared and Key_Shared subscription types)
（2）individually  (Exclusive and Failover subscription types).

question2：
but not have about "negatively acknowledged cumulatively" api，only have "negatively acknowledged individually" api(Consumer.negativeAcknowledeg)

i have 2 question 
thank you jun !

**Comment by labuladong** [comment]
You can try `negativeAcknowledge` in `Failover/Exclusive` subscription types. All the messages the consumer didn't acknowledge will be redelivered, this behavior could be negatively acknowledged cumulatively. But there do have some discussions about the redelivery behavior on `Failover/Exclusive` subscription types, see https://lists.apache.org/thread/vdwhg5wrpjc8x41oqwxrsdhx7dmyzt3j

**Comment by startjava** [comment]
but i don't know what mean "negatively acknowledged cumulatively"。

**Comment by momo-jun** [comment]
> but i don't know what mean "negatively acknowledged cumulatively"。

You can understand it as "Cumulative NACK" - the negative acknowledgement is performed cumulatively, that is, consumers only negatively acknowledge the last message they receive.

**Comment by tisonkun** [comment]
This is about knowledge sharing. Not an issue to resolve.
